### DISCUSSION TITLE
Cryengine 2.0:  Rough plans and solicitation for input

### DISCUSSION THREAD

**Discussion by Markemp** [Markemp/Cryengine-Converter#158] [General]
![Untitled Project](https://user-images.githubusercontent.com/12057128/232261530-ef7527ba-eeeb-4c5a-b327-0fa5eafc2e41.gif)

Well, the animation work (can't thank you enough @Soreepeong!)  seems to chugging along, so now it's time to make some decisions.  Here is my rough plans for what to do next, and I'd like to get the community's input.

Collada just doesn't seem like a great way forward anymore.  Thanks to Soreepeong's implementation of the glTF renderer, I'm not sure Collada has much of a place anymore.  Much like Waveform (.obj) files, it had its time.  And since Blender's implementation of USD is still a work in progress, I'm thinking that glTF is probably the best way forward for at least the intermediate future.  So here is what I'm thinking of doing for a 2.0 release:

- Change the default behavior to render to `glb`.  The other option is `gltf`, but I'm not sure of the differences on these yet.  Does glTF/glB add the texture files into the file too?
- Provide a way to have cgf-converter be aware (and be able to work on) the .pak files without having to extract everything.  I'm not really sure of how this will work (how do you find a file without traversing a directory?  Will a UI be required?  What do we do with games that have encryption, such as SC?).  I'm not even sure if this is a desired feature.  It will save the step of having to extract everything, but is that a compelling feature?
- Update **Cryengine Importer**, since that currently only works on Collada files.  It's something I have been needing to do for a while, along with making new videos on how to use it.  That includes how to use these fancy animations in the NLA!  (... as soon as I figure that out.  Blender has changed a lot!)
- For console based games, add a switch to the command line `-isConsoleGame`, so that the big endian processing takes place properly.  I don't see a really good way to identify a console vs pc game based on the game files themselves.  There is code that is in theory detects it, but I'm not really sure how (or if) that is working.  Even for console games, the header info and chunk tables (and chunk headers) are in little endian format.  It's just when the data is being read does this change. 😒
- I think I need to do some performance improvements.  It's taking a bit longer than I like to process files; far longer than it should. 

Any other thoughts on what to do for a 2.0 release?  It's going to take a bit of time, but hopefully I can get everything done by this summer.

**Comment by Kjasi** [comment]
I agree that Collada's time has come and gone, and with FBX being closed source, using an open source alternative such as USD or glTF sounds great! (Even better that glTF is already implemented, thanks to @Soreepeong's work!)

My thoughts on your points:
* `glB` is the binary implementation, while `glTF` is the ascii version. It's the difference between Maya's `.ma` and `.mb`. `glB` would be smaller, while `glTF` would be human-readable. My personal preference is for `glTF` to be the default, with `glB` as a fallback/optional format.
* Since P4K files seem to differ between games (and will even more so, with the upcoming SC P4K format rewrite) it might be best to NOT include a decompression system. It'd be just another thing to maintain, and there's already a lot of other cryengine decompression systems that people can already access. That said, having a system where people can identify their decompression software (unp4k, for example) and have the converter use that with hidden command-lines, might be a viable alternative to writing your own solution, and won't require you to maintain p4k formats it for multiple games.
* Having an updated CryEngine Importer would be nice, but I don't think it should be bundled with CGF-Converter. It should be it's own, separate project, because not everyone needs it. (StarFab users, for example, don't need it, they just need CGF-Converter.)
* Is the program multi-threaded? I know the reading has to be linear, but what about the processing of CGF files? Would that help conversion speed?
* I would love to see an optional GUI for CGF Converter for 2.0. One where you can preview the file before exporting would be nice!

**Comment by Soreepeong** [comment]
glb file is essentially a concatenated file of gltf and bin files, but with length information stored. I found myself ending up using glb exclusively once I got the textual part of the format working.

Also, it would be nice to have .editorconfig so that we won't have to "fight" automatic formating to keep the code format consistent!

**Comment by Markemp** [comment]
Thanks for the input!  Here's a AC list that I'll add to a 2.0 issue:

- Set default export to `.glb`.
- Add a new option `-isConsoleGame` to specify game files are from a console game to use big-endian reading for game data files.
- Update all integration tests.
- Duplicate existing integration tests for `.glb` versions.
- Add an .editorconfig.
- Increment version to 2.0.

My Cryengine Importer project will get its own set of issues to deal with the upgrade, since at the end of the day I need to be able to make mechs dance, otherwise what is even the point of all this! 🤣

**Reply by Dizcordum** [reply]
.caf format will be present in convertion options?

**Comment by Markemp** [comment]
Well, this is a bit more of a challenge than I expected.  With Collada, it was relatively easy to go from Cryengine -> Collada -> Blender, since everything was Z axis up.  With gltf, it only knows Y axis up, so I have to convert the Cryengine data to that axis when I write it, and then hope the gltf importer from Khronos handles it from there.  But unfortunately when I import something, it wants to use -Y as the forward direction instead of Y. 

That may not be a problem, but it probably will be when dealing with prefabs.  I'm guessing everything will be rotated around facing the wrong direction.  I think I'll just live with -Y being front for gltf models and when working on the prefab importers, just rotate them around there.  I still need to verify that child geometry is placed correctly, add VertsUvs, and double check vertex Colors since those seem a bit weird too.  But making progress at least!

**Reply by Soreepeong** [reply]
Take a look at SwapAxes.cs and see if you can get things right there!

**Reply by Markemp** [reply]
Oh nice!  I was wondering how all the animations and armatures were working!  I wrote a couple of dumb extensions for `Quaternion` and `Vector3` that really didn't seem to work properly, so will check out this class and see if it works a bit better! 👍🏽

**Reply by Soreepeong** [reply]
Honestly that class still doesn't make much sense for me else than axis changing for position; I almost was trying all the possible permutation of axes. Hope you get better mileage than me!

**Reply by Markemp** [reply]
Not having too much luck so far.  Feels like such a needle in a haystack.
![image](https://github.com/Markemp/Cryengine-Converter/assets/12057128/b951430c-954e-4358-8b7d-f21d2c17e643)

One of these 3 hula girls is from the Collada import (rotated 180 around Z to face the same direction), and the other 2 variations with using `SwapAxis` and just reordering X, -Z, Y.  Very close, but just a tiny bit off between the 3 attempts.  It gets very noticeable on the `green_fern_bush_a` game object from the Cryengine game engine, when the distances get bigger.

I'll keep poking around.  Having known good Collada values to compare against helps a lot at least.

**Comment by Markemp** [comment]
![image](https://github.com/Markemp/Cryengine-Converter/assets/12057128/9d01a6bb-e04a-479e-b9d2-d9392bc4b236)

Got the star citizen models importing correctly.  Everything is still flipped with -Y being forward, but I think I can fix this by just taking the negative of it when it's first read.  Hopefully that won't break animations too badly.

**Comment by Markemp** [comment]
![image](https://github.com/Markemp/Cryengine-Converter/assets/12057128/84464e49-6d32-4a8b-8b6b-349e28c7ea38)

Got the rotations fixed for all child nodes, and positive Y is forward again. I'm assuming this will break animations, but need to work on the Ivo files first, then will get that going next.  Making progress!

**Reply by Dizcordum** [reply]
Great!

**Comment by Markemp** [comment]
I think I found a way to combine the split DDS texture files for the newer game models so they go straight into the `.bin`/`.glb` file.  No more having to process them separately, and I can even do it in a non-destructive manner so that the original game files don't get modified. :)

I gotta stop getting sidetracked though. 😳

**Reply by Soreepeong** [reply]
CryEngine and glTF have different uses for some of texture channel layouts though; did you find a solution for that too? MSFT DDS extension does exist but I'm not sute if that can be used meaningfully here.

**Reply by Markemp** [reply]
Haven't looked at that yet, but will take a peek.  It would be really nice to get accurate materials right as part of the conversion, but I'm worried that texture files are studio dependent and it may be hard to get a general converter for them.  For example, the normal files for MWO don't use the regular red and blue channels, but instead use blue and green.  Why?  I don't know, but it means that normal files need custom nodes to work properly for those models.  🤷

**Comment by Markemp** [comment]
Ok, weird one here.  For some models, the Normal datastream has the normals as a 4 byte structure.  For example: `00 80 FF 1F`.

I'm pretty sure this is supposed to be a `Vector3` of some type, but not exactly sure how they are converting 4 bytes to 3 floats.  Maybe signed byte / 128?  Anyone have any ideas?

**Reply by Kjasi** [reply]
Without more context, I can only guess, but maybe it's a Quat, with each axis compressed into a single byte? So X would be 0/255, Y is 80/255, Z is FF/255, and w is 1F/255? (Alternatively, WXYZ...) You would then convert the Quat into a Vec3.

**Reply by Soreepeong** [reply]
The existence of FF and 80 makes it difficult to take the values as either SNORM, UNORM, or 2's compliment. In either case the values involve either 1 or -1, and it makes the Vec3 or Quat/Vec4 so not normalized, which I think would be unlikely.

**Comment by Markemp** [comment]
Oh wow, just discovered the glTF plugin for VS Code.  It's so nice!  It can validate the gltf and glb files, and even has a 3d viewer built in!  This is so much nicer for verifying models are converting properly than looking at them in notepad++ and then loading into Blender.  🤤

**Comment by Markemp** [comment]
Starting to get textures figured out.  Star Citizen doesn't use names, but rather texture slots.  I'm trying to figure out what each of them are.
        TexSlot1,
        TexSlot2,  // Normal
        TexSlot3,
        TexSlot4,  // Specular
        TexSlot9,  // Diffuse
        TexSlot10, // Specular
        TexSlot12, // Blend
I'm assuming that the missing numbers have their own definitions as well, but haven't run across them yet.

![image](https://github.com/Markemp/Cryengine-Converter/assets/12057128/cb818e44-30de-496b-9cfd-bef4ddf78384)

**Reply by Kjasi** [reply]
The slot numbers will change depending on the shader setup. For the Illum shader, your texture slots are a good start, but you have to understand there's a LOT of options. For one, if the StringGenMask contains BLENDLAYER, then the material uses a two layer setup.
Layer 1 is usually Slot 1, 2, 4, 6, and 8, with Layer 2 using 3, 9, 10, 11, and 13.

For the Illum shader, TexSlot 1 and 9 are diffuse/color maps. 2 and 3 are the normal maps, 4 and 10 are specular, 6 and 13 are Detail Mapping textures. (modifies diffuse, specular, and normals, iirc) and 8 and 11 are displacement textures used for Parallax Occlusion. Slot 5 seems to be a reflection map, usually grabbed from the environment at runtime. Slot 12 is the blendmap that somehow mixes Layer 1 and 2. (My testing seems to indicate it's not just a straight lerp...) And let's not get into the mess that is the Vertex Colors, which seems to be used all throughout all the materials... (And we need all 4 vertex colors for the materials to look right... This sucks, because Collada only supports 3-channel vertex colors...)

Other shaders can use a different slot assignment, such as GlassPBR using Slot 6 for a Gloss Map. And the HardSurface shader usually doesn't use texture slots at all; They mix other materials (which are usually using the Layer shader) using a Primary/Wear setup, but HardSurface will sometimes use one (or more) texture slots that would over-ride texture slots inside the referenced materials. (Slot numbers for HardSurface seem to follow the Illum's layer setup, but are applied to proper slots inside the referenced materials. So, for example, Slot 3 would apply over the normal inside the wear's material.)

I should mention that, in my research, it seems that Slot 1 is always the base Diffuse Color, and Slot 2 is always the base Normal map.

**Reply by Soreepeong** [reply]
Sounds like even though I've heard that SC has diverged a lot from CE, these definitions still mostly applies? Does that game get shipped with such definition files like this? (Also, I finally figured out how bsg files in RoL are used.)

![image](https://github.com/Markemp/Cryengine-Converter/assets/3614868/c5a6370c-2ee4-45f1-b213-f422fbd20f1a)

**Reply by Kjasi** [reply]
Those files exist in SC, but their description isn't really helpful in identifying the texslots... Example: (From the Illum.ext file)
```
Property
{
  Name = %DETAIL_MAPPING
  Mask = 0x4000
  Property    (Detail mapping)
  Description (Use detail mapping)
  DependencyReset = $TEX_Detail
}

Property
{
  Name = %VERTCOLORS
  Mask = 0x400000
  Property    (Vertex Colors)
  Description (Use vertex colors)
}

Property
{
  Name = %PARALLAX_OCCLUSION_MAPPING
  Mask = 0x8000000
  Property    (Parallax occlusion mapping)
  Description (Use parallax occlusion mapping (requires height map (_displ)))
  DependencyReset = $TEX_Normal
}

Property
{
  Name = %DISPLACEMENT_MAPPING
  Mask = 0x10000000
  Property    (Displacement mapping)
  Description (Use displacement mapping (requires height map (_displ)))
}
```

**Reply by Markemp** [reply]
What a mess.  And somehow we need to come up with a neat and tidy algorithm so all these cases work in Blender, or at least the gltf file can properly reflect these material settings. :(

I did solve some normal issues for older game files.  The blue channel for the MWO `.ddn` files were set to all 0, which caused the images to have a yellow hue.  By setting them all to 1, they look like proper normal maps, and lighting actually works like it is supposed to again.  If it's a newer `.ddna` file, it won't change the blue values to `0xFF`.  

The Cryengine docs mention that the normal should be normalized if possible, but it's not required:
https://docs.cryengine.com/display/SDKDOC4/Tangent+Space+Normal+Mapping
I'm just going to set the values to `0xFF` unless someone thinks there is a good reason to do the calculation.

One other drawback to the gltf files:  I'm including the texture files in the converted file, but this can make the gltf files huge.  For example, the Avenger ship in SC is over 100MB when converted.  That isn't the end of the world, but when creating a scene from a prefab or loading in a level, there isn't really a way to reuse these textures.  The resulting scenes could be unworkable.  Even importing a mech may be tough for Blender to handle, since there are dozens of parts and they all reuse the same 4 or 5 textures.  Anyone know of a way to share texture libraries in gltf files? :(

**Reply by Soreepeong** [reply]
Since normal maps can formed from two channels, it may be the case they dropped the blue channel, but glTF rendering programs will expect all three channels to be present, and they will normalize by scaling each value, rather than `b=sqrt(1-pow(r,2)-pow(g,2))`. We do want the values restored on exporting.

And no, you cannot reuse textures across multiple glTF files. The closest you can do is to export a gltf file containing all the objects you want to use, so that the file does not contain duplicate textures.

**Reply by Kjasi** [reply]
In regards to normal maps; I've run into instances where some games developers decide to omit the blue channel, and recalculate it's values at runtime, and others will invert the channel, since the blue channel is mostly white anyways. Either way, they're storing a mostly black channel, rather than a mostly white channel, which will eventually save them quite a bit of texture space. (None of the ones I've found were CryEngine games, but that may be happening with MWO...)

Does the glTF NEED to contain the textures inside it? Couldn't it just reference the textures relative to the source input file, and the extracted textures are placed next to their origin files? That would make it library friendly...

**Reply by Soreepeong** [reply]
It can refer to textures specified by relative paths, but importing it on Blender would always add files as new textures afaik.

**Reply by Kjasi** [reply]
Yeah, it would, but a simple script can condense texture references to a single reference.

**Reply by Markemp** [reply]
I think that is how it's going to have to be done, with a script.  When a gltf model is imported into a scene as part of a script, it'll need to check for duplicate textures with that name, point the material to use that, and then delete the duplicate texture.  A bit annoying, but I think that will be a workable solution.

**Reply by vmxeo** [reply]
FYI - I've been hacking at much of this in Starfab, however texture slots are not consistent across all shader types. There's additional work that needs to be done with the textures under the hood. Not as simple as plugging in each texture to a PBR channel.

**Comment by Dizcordum** [comment]
HI! Any update on new release?
I hyped for that animations export support

**Reply by Markemp** [reply]
I've been stuck for the last month trying to get the armatures working in gltf files for skins.  I just can't get the orientation right.  It's driving me nuts. :(

![image](https://github.com/Markemp/Cryengine-Converter/assets/12057128/cf0970f3-498f-4b3c-8bd6-699d6f00f538)

**Reply by Dizcordum** [reply]
Is it something root bone orientation? or the animation itself?
Need any help?

**Reply by Markemp** [reply]
Yeah, I wouldn't mind a pair of fresh eyes on this.  I think it's in the `BaseGltfRenderer.Geometry.cs` file, starting around line 229 (the `foreach` statement for all the bones) in my branch `feature/159-2.0`.  You can compare how it's done in the `ColladaModelRenderer.cs` file as well, in the `CreateJointNode()` method.  The matrix is pre-calculated in the bone's `LocalTransform` variable.  I just use that directly in Collada, I just can't find the trick to rotating it around the Z axis 180 degrees.

Once this is lined up properly (it's in Edit mode in the screenshot), I'll work on the bind pose matrix (so the skinning has the proper weights), and then finally to animations.

It's easy in Collada because it's all Z up coordinates along with Blender, but gltf is Y up, so I have to go from Z up to Y up and back.  I'm attaching the chicken collada file for reference.  The model is from Armored Warfare under `objects/characters/animals/birds/chicken`. Can't attach them to here directly, but let me know if you need it. 

[chicken.dae.txt](https://github.com/Markemp/Cryengine-Converter/files/12678838/chicken.dae.txt)

**Reply by giniedp** [reply]
@Markemp i'm trying to get access to the chicken model. Yet i cant find how to extract Armored Warfare files. If using Quickbms, you have a bms script for me?

**Comment by Markemp** [comment]
Eh... maybe Collada is the way to go after all... 🤔

**Reply by giniedp** [reply]
its been a while for me to convert between file formats, maybe i'm not seeing the whole picture, but a thought on my mind is: couldnt you just use data as is, and then just add a root transform node to bring to Y up?

**Reply by Markemp** [reply]
I've tried that.  Or at least I think I've tried it.  So many different iterations. :(  But I'm making decent progress on the Collada implementation, so if I can get that to work I'll just go back to that as the default.  That also solves the issue with the image files being included in the binary data, which causes the converted files to be rather big (300+ MB just for the SC Avenger ship).  Hopefully I'll get the x905 version of the Controllers implemented for the Collada converter, and then can work on the older styles.

**Reply by giniedp** [reply]
i would totally give it another shot. Might try for myself next week. Actually i would be fine with the converter extracting without transformation. At least at first.

**Reply by giniedp** [reply]
if you are talking about embedding images into binary `.glb` format? then this is actually not needed. While you "can" embed images into the binary buffer, you dont "have" to do it. Since the `.glb` is only a container for the `.gltf` json and the binary data. And the `.gltf` can reference external images, regardles whether it is contained in the `.glb` or not. Or do you mean some other problem?

**Reply by Markemp** [reply]
Oh, that's exactly what I meant.  How do you reference an external image with glTF?  I'd rather not include texture files into the output if at all possible.

**Reply by giniedp** [reply]
The image object can have a `uri` property which may be a data url, or an absolute url or a relative path
- https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#images
- https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#uris

You actually have modeled that field
https://github.com/Markemp/Cryengine-Converter/blob/28190fcf1bfd8f111a3be89e17fab72bd2f707fb/CgfConverter/Renderers/Gltf/Models/GltfImage.cs#L10

I'd also prefer the images being referenced externally. At least in my case i'll do postprocessing anyways and convert materials, add gltf plugins, optimize images with webp or ktx compression
- https://www.khronos.org/ktx/
- https://github.com/KhronosGroup/KTX-Software

Also i'd like to have an option to either output binary or json (`.glb` or `.gltf`) but, lets fix the converter first, and then talk options

**Reply by Markemp** [reply]
Here's the `chicken.chr` file.  
https://1drv.ms/u/s!Ahz_oM1_DqLRkOoR77MnXINpfm3uYA?e=ARpiCo

The externally referenced images is really important for dealing with the prefabs.  The blend files would get insane otherwise, so thainks for finding that.

I'll push up my latest changes in the `feature/159-2.0` branch.  The relevant code is in `BaseGltfRenderer.Geometry.cs`.

There already is the option to select `-glb` or `-gltf` too.  glTF is nice because you can see it in a text editor.

**Reply by giniedp** [reply]
So, i have disabled all the swap functions (made them return the input parameter) and added an artificial root node to the scene, with the rotation from Z UP to Y UP. Results:

![image](https://github.com/Markemp/Cryengine-Converter/assets/225238/f4a7fbfb-3f15-486d-bdcc-2238c4ac89d4)

Looks good, I think, unless I missed something.

```cs
    public int Render()
    {
        // Create the root object
        Reset("Scene");

        var rootIndex = _gltfRoot.Nodes.Count;
        var root = new GltfNode()
        {
            Name = "Y_UP",
            Rotation = Quaternion.CreateFromYawPitchRoll((float)Math.PI, -(float)Math.PI / 2.0f, 0).ToGltfList(),
            //Rotation = Quaternion.CreateFromYawPitchRoll(0, 0, 0).ToGltfList(),
            Translation = Vector3.Zero.ToGltfList(),
            Scale = Vector3.One.ToGltfList()
        };
        _gltfRoot.Nodes.Add(root);

        CurrentScene.Nodes.Add(rootIndex);

        // For each root node in the crydata, add to the scene nodes.
        foreach (var cryNode in _cryData.Models[0].NodeMap.Values.Where(a => a.ParentNodeID == ~0).ToList())
        {
            // CurrentScene.Nodes has the index for the nodes in GltfRoot.Nodes
            root.Children.Add(CreateGltfNode(cryNode));
        }

        Save(_cryData.InputFile);
        return 1;
    }
```

**Reply by Markemp** [reply]
Nice, looks right!  Is this based off of the `master` branch or my feature branch?

**Reply by giniedp** [reply]
the feature branch

**Reply by Markemp** [reply]
I think I got this to work.  Had to make a few changes since the scene and node layout changed a bit (bones aren't supposed to be children of the nodes, so did some refactoring there).  But here's the MWO adder armature doing it's thing!

![Adder animation](https://github.com/Markemp/Cryengine-Converter/assets/12057128/5dd8221b-6e50-48ca-b9e6-e419036cb5e8)

Going to do some more testing, but this is looking very promising!

**Reply by Markemp** [reply]
I'm still not sure this is right.  The geometry when imported into Blender is off from the Collada imports, for models like the MWO hula girl and the Cryengine green fern.  And I'm not sure why.  Is it reasonable to think that, if I take a Z up, Y forward model from Cryengine, convert it without changing the transform to Y Up, and then the glTF importer brings it back in to Blender would undo those changes, but it just doesn't work out right.

Maybe the glTF inporter in Blender isn't right.  I dunno.  I'll just keep going down this path for now, and hopefully I won't run into too many things that are broken.

**Reply by Markemp** [reply]
Yeah, this method won't work. :(  Ugh, what a pain.  Back to the drawing board.
![image](https://github.com/Markemp/Cryengine-Converter/assets/12057128/20db4531-7672-40e2-a807-ae901c1cafd8)

**Reply by giniedp** [reply]
@Markemp can you send me the model? I'd take a look during this week.

**Reply by Markemp** [reply]
https://1drv.ms/u/s!Ahz_oM1_DqLRkOsqfvA4akVzYeAOtg?e=4T1VND

The Z-up, Y-forward result is important since all the prefabs will be oriented that way as well.  The axis swap stuff works fine on plain objects, but it's just failing on the armatures, like the chicken model from earlier.  Solving the armature rotation thing is the real trick.  Well... probably until the next trick of sorting out the animation values.  But that's a problem for later.

**Comment by Markemp** [comment]
Added an optional argument `-embedTextures` so that glTF can embed the texture files.  Won't do anything for Collada, but should make converting to glTF a lot faster by default.

**Comment by Markemp** [comment]
I think I'm going to put the animation work on hold and just finish up a 1.6 update with all the other backlog stuff that has been festering, such as the new Ivo format for SC and better material handling.  I'm just spinning my wheels at this point.  I've even run Blender from source and traced through the code to try to get a better understanding of what's going on, and... I think I just have a mental block.  Sorry everyone.  I'll get back on the animation stuff after a little break, but will start the 1.6 work this weekend.

**Comment by Baconator650** [comment]
No need to apologize, animation is its tricky beast.  

As someone who represents that SC community, we are very excited about the new IVO format being fixed, so we can have our landing gear. Especially with the new ships that have recently been released, and what appears to be noodles. 

As one of the few people doing animation with SC assets (via the Dymek Engine and non-animation files) I feel like setting aside the CGF-Converter animation support will not slow down the progress of the few who are experimenting with it (honestly I think I might be the only one). This especially being the case as StarFab currently does not access to the animation database .

**Reply by Markemp** [reply]
Ok, I feel really dumb, but... I just installed 3.21 to get the latest files and can't find the `.pak` files anywhere.  I installed to `d:\starcitizen` and can see a `data` subdirectory, but it's empty except for a `lib` subdir.  Where are all the data files?

**Reply by Baconator650** [reply]
No problem, the .p4k (remember SC names their ,pak files with a "4", simply renaming the extension .pak usually works). Files location is usually ( \Program Files\Roberts Space Industries\StarCitizen\LIVE)

![image](https://github.com/Markemp/Cryengine-Converter/assets/78010776/1fc6af1d-f3c3-463b-b4fb-2e505e6e4569)

If that's not there, I'd try a file verification via the launcher and then make sure you can load arena commander or such. I had similar issue with one of the testing builds

**Reply by Markemp** [reply]
Omg no wonder I couldn't find it. It was right in front of me. 🤦‍♂️

Thanks for the assist!

**Reply by Baconator650** [reply]
It's a pleasure

**Reply by Baconator650** [reply]
If you can, could you add an argument or go back to the old naming convention of "file name"+_mtl_+"material name" to be compatible with the StarFab Blender Plugins and 3DS Max workflows

**Comment by Markemp** [comment]
I think I finally solved material issues.

![image](https://github.com/Markemp/Cryengine-Converter/assets/12057128/5cab1dab-a01d-4e0e-99f8-1faad2d8bd7e)

It does require you to pass in the mtl file for the object in some cases.  But at least I was able to delete a lot of code related to "if this weird ass edge case, then try to hack in this material".  😌 

I think normals are all messed up.  And always have been.

**Reply by Baconator650** [reply]
Keep up the good work! Some of us are already experimenting with the changes by compiling a proto-type version with the changes in the repository.

Don't forget the request from the SC community for the material naming format as  "mtl file name"+mtl+"material name"  [or option]. Its a small thing, but a huge help.

**Reply by Soreepeong** [reply]
Just noting that when I was exporting from Rise of Lyric, I had to use CryEngine Binormal for GLTF Tangent, and conversely had to calculate CryEngine Tangent from GLTF Normal + Tangent when importing from GLTF into CryEngine model; otherwise, left half (or right half?) would have strange shading in-game.

**Reply by Markemp** [reply]
Yup, planning on getting the mtl naming updated here shortly.  Just making sure that the glTF exports can handle the new way we're handling materials.  Bonus is that instead of processing the material files for every node chunk, it just processes the mtl file once to create the materials.

Soreepeong, where did you get the Rise of Lyric files?  I don't have any good integration tests for those models, especially related to the pack file systems stuff that you added.  I'm really nervous about breaking exports, and would love to add the game to my depot.  I have a couple of models that I can test with, but since the converter is pretty dependent on maintaining the game data structure, it still makes me nervous to make changes that could affect those exports.

**Reply by Soreepeong** [reply]
https://github.com/Soreepeong/SynergyTools I've made a full import/export tool for RoL; it can export a model to a file and import back seemingly without losing details or animations, and apparently works fine enough to even import from models exported from FFXIV. If you don't have the game files available with you, Internet Archive got the files; or I also can send you some exported files.

**Reply by Markemp** [reply]
@Soreepeong I feel like I'm messing up the glTF stuff with this 1.6 update.  I changed how materials work (instead of loading from each node chunk, the materials are created from the mtl file right away and added to the root `CryEngine` obect).  I tried to refactor the materials to create the glTF material and textures right from that material, but dealing with the indices and accessors it a challenge.  Any chance you feel like taking a look at the latest branch and seeing if you can get materials to work with the new materials format?

@Baconator650, would "<mtl file name>-<mtl material name>" be ok as well?  I'm pretty sure the dash isn't used in any mtl file or material name, so would be a good character to split on.

**Reply by Baconator650** [reply]
All the scripts and workflows are built around underscores "_" for the split and also for any spaces/blanks in the material name (we had an issue way back then with the spaces/blanks in the material names and the solution was to replace it with an underscore, as a result, the scripts built around this).

Format used is:

MtlName=.mtl filename
Mat=Material Name (replace " " with "-")

Final Material Name=MtlName + " [underscore]mtl[underscore]" + Mat

![image](https://github.com/Markemp/Cryengine-Converter/assets/78010776/aa7deafd-8fd5-466a-b41a-9187798be975)


Worst case scenario if we replace the underscore with a hyphen; I am sure the vmxeo could rewrite the blender scripts for to accommodated the change. And I can write a prehandling script for Max that would rename it to the proper format.

**Reply by Soreepeong** [reply]
I'm doing that right now; hopefully it wouldn't take too long!

**Comment by serpicos** [comment]
any of you guys triend the tool extracting aion models?

**Reply by Markemp** [reply]
It should work for that game as well.  I'll get it downloaded and make sure it generally works for those files.  They always could implement some weird chunks that haven't been implemented yet.

**Reply by serpicos** [reply]
It would be amazing if you can check it, i'll try also even if i'm nop super expert, the version of the game is Aion that comes at gameforge website
Telling this because there is also another version called aion classic, delivere by the ncsoft

**Reply by Markemp** [reply]
I found the NCSoft version.  Are there multiple Aion games?

**Reply by serpicos** [reply]
the one you found is a newer version, that follow the first steps of the games, a bit like classic wow servers
the standard and full version can be found there sir
https://gameforge.com/en-GB/play/aion

**Reply by serpicos** [reply]
Managed to extract the pak files from the game (the one from gameforge website)
Tried with a model that should be a mount
but getting this error sadly

********************************************************************************
There was an error rendering D:\maghette\aionmount\KuillusRIDER03.cgf

Unsupported FileS ignature NCAion

   at CgfConverter.CryEngineCore.Model.ReadFileHeader(BinaryReader b)
   at CgfConverter.CryEngineCore.Model.Load(String fileName)
   at CgfConverter.CryEngine.ProcessCryengineFiles()
   at CgfConverterConsole.Program.Main(String[] args)
********************************************************************************

let me know if you need the files for an investigation

**Reply by Markemp** [reply]
I feel like I've looked at these Aion files in the past, and despite the file extension of `.cgf` they aren't actually Cryengine games. :(
